Improve build speed [manually] (forward-declarations/clean up includes/etc.)

Geometror opened the pull request:
This PR aims to improve compilation time by forward-declaring some classes whenever possible/when it makes sense and cleaning up/moving some include statements. There is still some more room to improve, but I think this should be done in subsequent PRs as this manual process of finding optimization opportunities can be very time consuming.

Changes:
- Add forward-declarations of `EditorFileDialog` and `EditorNode` to several classes. This is already done in some cases, but not consistently.
- Remove unused includes of `file_dialog.h`
- Remove forward-declaration of Container in `flow_container.h` as it was used as a base class which can cause problems. (could be moved to a separate PR)
- Optimize includes in `EditorNode` and sort forward declarations
- Make missing relative includes (primarily `editor` includes) absolute for consistency

Some benchmarks (as of Feb 11 2022):
[Time of rebuild. before -> this PR]

Change in progress_dialog.h
**28s -> 23s (~17% reduction)**

Change in dependency_editor.h
**25s -> 23s (~8% reduction)**

Change in editor_data.h
**3:55m -> 3:01m (~23% reduction)**

Change in file_dialog.h
**3:47m -> 21s (~91% reduction)**

Change in line_edit.h
**4:37m -> 3:50m (~17% reduction)**

Changed all files above:
**4:30m -> 3:56 (~13% reduction)**

Calinou commented:
This is a good start, nice work :slightly_smiling_face: 

Did you see https://github.com/godotengine/godot/issues/57569, by the way?

Geometror commented:
@Calinou Thats quite interesting, will see what I can do with that :)
I am currently trying to create a dependency graph of the godot codebase to find some more optimization opportunities. Does anyone know a good tool (Windows/Linux) to render large/complex graphs and analyze them? I already have the DOT data.

Calinou commented:
There's [include-what-you-use](https://include-what-you-use.org/) but I don't know how difficult it would be to run on a SCons codebase.

akien-mga commented:
> There's [include-what-you-use](https://include-what-you-use.org/) but I don't know how difficult it would be to run on a SCons-using codebase.

It's not too difficult, it just needs a wrapper script. I ran it a few years ago, but it's not trivial to make good use of it as it affects the whole codebase.

When using it, a good first step would be to try to sort includes in `core` only (and anything else that requires changes due to core include changes).

Geometror commented:
@Calinou @akien-mga 
As you said, getting it to work was straight-forward, however there are problems which makes it quite difficult to make good use of it:

- Sometimes it removes includes of generated headers
- platform specific includes are added (this can be easily fixed though)
- it fails at some complex templates (with specializations) which are declared across multiple files

Tried it in `core` but couldn't get it to work because of the last point above. It always fails at the GetTypeInfo template and I don't know why. If someone wants to have a look at it:
https://github.com/Geometror/godot/tree/iwyu-core-1

However, running it over the `editor` folder worked fine, although still requiring some manual work. See #57734 for some quick benchmarks.

Geometror commented:
Rebased, squashed and did some further cleanup, now I think its ready for review.
Added some benchmarks to my original message above, feedback is always welcome :)

akien-mga commented:
This looks excellent, I'll review and test asap.

akien-mga commented on core/debugger/remote_debugger.cpp:
```diff
@@ -38,6 +38,7 @@
 #include "core/input/input.h"
 #include "core/object/script_language.h"
 #include "core/os/os.h"
+#include "scene/main/node.h"
```
That looks like a rebase mistake, this was actually removed on purpose in 65837973051e2aea43019fb5c20ec4a1debf0bcb. `core` shouldn't depend on `scene`.

akien-mga commented on core/variant/typed_array.h:
```diff
@@ -31,8 +31,10 @@
 #ifndef TYPED_ARRAY_H
 #define TYPED_ARRAY_H
 
+#include "core/object/object.h"
```
Note for later (needs discussion), this seems needed only for `PropertyInfo` and related stuff, we might want to check if it would be useful to split property stuff out of `object.h` (e.g. `core/object/property_info.h`).

akien-mga commented on editor/dependency_editor.h:
```diff
@@ -31,8 +31,8 @@
 
-#include "editor_file_dialog.h"
 #include "scene/gui/dialogs.h"
+#include "scene/gui/item_list.h"
```
Could maybe be forward declared?

akien-mga replied:
Nevermind this and other comments about `dependency_editor.h`, I was a bit too thorough in my review initially before giving up in front of the sheer content - that's stuff for future PRs :)

akien-mga commented on editor/dependency_editor.h:
```diff
@@ -31,8 +31,8 @@
-#include "editor_file_dialog.h"
 #include "scene/gui/dialogs.h"
+#include "scene/gui/item_list.h"
 #include "scene/gui/tab_container.h"
```
Seems unused.

akien-mga commented on editor/dependency_editor.h:
```diff
@@ -31,8 +31,8 @@
 #include "scene/gui/dialogs.h"
+#include "scene/gui/item_list.h"
 #include "scene/gui/tab_container.h"
 #include "scene/gui/tree.h"
```
Might be able to move to the .cpp.

akien-mga commented on editor/editor_node.cpp:
```diff
@@ -47,6 +47,11 @@
 #include "core/string/print_string.h"
 #include "core/string/translation.h"
 #include "core/version.h"
+#include "editor/editor_data.h"
+#include "editor/editor_folding.h"
+#include "editor/editor_run.h"
+#include "editor/editor_toaster.h"
+#include "editor/scene_tree_dock.h"
```
This class seems to have `editor/` includes intentionally separated from the rest further below.

akien-mga commented on editor/editor_node.h:
```diff
@@ -32,16 +32,13 @@
 #include "editor/editor_native_shader_source_visualizer.h"
 #include "editor/editor_run.h"
-#include "editor/editor_toaster.h"
+
```
Unnecessary separation line.

akien-mga commented on editor/editor_toaster.cpp:
```diff
@@ -28,13 +28,13 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
+#include "editor/editor_toaster.h"
```
```suggestion
#include "editor_toaster.h"
```
For the header corresponding to this `.cpp` we don't use an absolute path normally.

akien-mga commented on editor/editor_toaster.h:
```diff
@@ -33,6 +33,7 @@
 
 #include "scene/gui/box_container.h"
 #include "scene/gui/button.h"
+#include "scene/gui/panel_container.h"
```
It's also in the .cpp, so the .cpp one can be removed (or the header one can be forward declared).

akien-mga commented on editor/import/scene_import_settings.cpp:
```diff
@@ -29,12 +29,15 @@
+#include "scene/animation/animation_player.h"
 #include "scene/resources/importer_mesh.h"
 #include "scene/resources/surface_tool.h"
-
```
Please keep this :)

akien-mga commented on editor/import/scene_import_settings.cpp:
```diff
@@ -29,12 +29,15 @@
 /*************************************************************************/
 
 #include "scene_import_settings.h"
+#include "editor/editor_file_dialog.h"
+#include "editor/editor_file_system.h"
```
While at it:
```suggestion
#include "scene_import_settings.h"

#include "editor/editor_file_dialog.h"
#include "editor/editor_file_system.h"
```

akien-mga commented on editor/plugins/cpu_particles_3d_editor_plugin.h:
```diff
@@ -34,6 +34,7 @@
 #include "editor/plugins/gpu_particles_3d_editor_plugin.h"
 #include "scene/3d/cpu_particles_3d.h"
 
+class EditorNode;
 class CPUParticles3DEditor : public GPUParticles3DEditorBase {
```
```suggestion
class EditorNode;

class CPUParticles3DEditor : public GPUParticles3DEditorBase {
```

akien-mga commented on editor/plugins/gpu_particles_3d_editor_plugin.h:
```diff
@@ -31,11 +31,12 @@
 #include "scene/3d/gpu_particles_3d.h"
 #include "scene/gui/spin_box.h"
 
+class EditorNode;
+class SceneTreeDialog;
 class GPUParticles3DEditorBase : public Control {
```
```suggestion
class EditorNode;
class SceneTreeDialog;

class GPUParticles3DEditorBase : public Control {
```

akien-mga commented on editor/plugins/gpu_particles_collision_sdf_editor_plugin.cpp:
```diff
@@ -30,6 +30,8 @@
 
 #include "gpu_particles_collision_sdf_editor_plugin.h"
 
+#include "editor/editor_file_dialog.h"
+#include "editor/editor_node.h"
 void GPUParticlesCollisionSDF3DEditorPlugin::_bake() {
```
```suggestion
#include "editor/editor_file_dialog.h"
#include "editor/editor_node.h"

void GPUParticlesCollisionSDF3DEditorPlugin::_bake() {
```

akien-mga commented on editor/plugins/lightmap_gi_editor_plugin.h:
```diff
@@ -31,11 +31,13 @@
 #include "scene/3d/lightmap_gi.h"
 #include "scene/resources/material.h"
 
+class EditorNode;
+struct EditorProgress;
+class EditorFileDialog;
 class LightmapGIEditorPlugin : public EditorPlugin {
```
```suggestion
class EditorNode;
struct EditorProgress;
class EditorFileDialog;

class LightmapGIEditorPlugin : public EditorPlugin {
```

akien-mga commented on editor/plugins/occluder_instance_3d_editor_plugin.h:
```diff
@@ -31,11 +31,13 @@
 #include "scene/3d/occluder_instance_3d.h"
 #include "scene/resources/material.h"
 
+class EditorNode;
+struct EditorProgress;
+class EditorFileDialog;
 class OccluderInstance3DEditorPlugin : public EditorPlugin {
```
```suggestion
class EditorNode;
struct EditorProgress;
class EditorFileDialog;

class OccluderInstance3DEditorPlugin : public EditorPlugin {
```

akien-mga commented on editor/plugins/polygon_2d_editor_plugin.h:
```diff
@@ -35,7 +35,9 @@
 class ViewPanner;
 class ScrollContainer;
-
+class Panel;
+class HSlider;
+class SpinBox;
 class Polygon2DEditor : public AbstractPolygon2DEditor {
```
```suggestion
class Panel;
class HSlider;
class SpinBox;

class Polygon2DEditor : public AbstractPolygon2DEditor {
```

akien-mga commented on editor/plugins/resource_preloader_editor_plugin.h:
```diff
@@ -31,13 +31,14 @@
 #include "scene/gui/tree.h"
 #include "scene/main/resource_preloader.h"
 
+class EditorNode;
+struct EditorProgress;
+class EditorFileDialog;
 class ResourcePreloaderEditor : public PanelContainer {
```
```suggestion
class EditorNode;
struct EditorProgress;
class EditorFileDialog;

class ResourcePreloaderEditor : public PanelContainer {
```

akien-mga commented on editor/plugins/script_editor_plugin.h:
```diff
@@ -47,6 +47,9 @@
 #include "scene/main/timer.h"
 #include "scene/resources/text_file.h"
 
+class EditorNode;
+struct EditorProgress;
+class EditorFileDialog;
 class EditorSyntaxHighlighter : public SyntaxHighlighter {
```
```suggestion
class EditorNode;
struct EditorProgress;
class EditorFileDialog;

class EditorSyntaxHighlighter : public SyntaxHighlighter {
```

akien-mga commented on editor/plugins/style_box_editor_plugin.h:
```diff
@@ -32,11 +32,12 @@
 #include "scene/gui/texture_rect.h"
 #include "scene/resources/style_box.h"
 
+class EditorNode;
 class StyleBoxPreview : public VBoxContainer {
```
```suggestion
class EditorNode;

class StyleBoxPreview : public VBoxContainer {
```

akien-mga commented on editor/plugins/theme_editor_plugin.h:
```diff
@@ -31,17 +31,21 @@
 
-#include "editor/editor_node.h"
-
+class EditorNode;
+class EditorFileDialog;
 class ThemeItemImportTree : public VBoxContainer {
```
```suggestion
class EditorNode;
class EditorFileDialog;

class ThemeItemImportTree : public VBoxContainer {
```

akien-mga commented on editor/plugins/theme_editor_preview.h:
```diff
@@ -32,25 +32,13 @@
 
 #include "scene/gui/box_container.h"
-#include "scene/gui/check_box.h"
-#include "scene/gui/check_button.h"
-#include "scene/gui/color_picker.h"
+#include "scene/gui/button.h"
 #include "scene/gui/color_rect.h"
-#include "scene/gui/label.h"
 #include "scene/gui/margin_container.h"
-#include "scene/gui/menu_button.h"
-#include "scene/gui/option_button.h"
-#include "scene/gui/panel.h"
-#include "scene/gui/progress_bar.h"
 #include "scene/gui/scroll_container.h"
-#include "scene/gui/separator.h"
-#include "scene/gui/spin_box.h"
-#include "scene/gui/tab_container.h"
-#include "scene/gui/text_edit.h"
-#include "scene/gui/tree.h"
```
Woooosh :D

akien-mga commented on editor/plugins/voxel_gi_editor_plugin.h:
```diff
@@ -31,11 +31,13 @@
 #include "scene/3d/voxel_gi.h"
 #include "scene/resources/material.h"
 
+class EditorNode;
+class EditorFileDialog;
+struct EditorProgress;
 class VoxelGIEditorPlugin : public EditorPlugin {
```
```suggestion
class EditorNode;
class EditorFileDialog;
struct EditorProgress;

class VoxelGIEditorPlugin : public EditorPlugin {
```

akien-mga commented on editor/scene_tree_editor.cpp:
```diff
@@ -31,7 +31,9 @@
 #include "scene_tree_editor.h"
 
 #include "core/object/message_queue.h"
+
```
Empty line can be removed.

akien-mga commented on modules/csg/csg_gizmos.cpp:
```diff
@@ -29,6 +29,7 @@
 /*************************************************************************/
 
 #include "csg_gizmos.h"
+#include "editor/editor_settings.h"
 #include "editor/plugins/node_3d_editor_plugin.h"
```
While at it:
```suggestion
#include "csg_gizmos.h"

#include "editor/editor_settings.h"
#include "editor/plugins/node_3d_editor_plugin.h"
```

akien-mga commented on modules/gltf/register_types.cpp:
```diff
@@ -48,6 +46,8 @@
 #include "gltf_spec_gloss.h"
 #include "gltf_state.h"
 #include "gltf_texture.h"
+#include "modules/gltf/editor_scene_exporter_gltf_plugin.h"
+#include "modules/gltf/editor_scene_importer_gltf.h"
```
In this case it's wrong, modules are made to be modular and should not hardcode their absolute path. It should still compile without change if the module was renamed or moved out of tree.

akien-mga commented on modules/gltf/editor_scene_exporter_gltf_plugin.h:
```diff
@@ -31,11 +31,11 @@
 #ifndef EDITOR_SCENE_EXPORTER_GLTF_PLUGIN_H
 #define EDITOR_SCENE_EXPORTER_GLTF_PLUGIN_H
 
+#include "editor_scene_importer_gltf.h"
```
That seems incorrect? Editor code only makes sense to include when `TOOLS_ENABLED` is defined.

aaronfranke replied:
Isn't this entire file (`editor_scene_exporter_gltf_plugin.h`) already only for editor situations, so it wouldn't matter if `editor_scene_importer_gltf.h` is always included in it?

akien-mga replied:
In theory yes, in practice no, the `register_types.cpp` for the gltf module doesn't properly put editor-specific includes within `TOOLS_ENABLED`. And since all `.cpp` files get globbed in the `SCsub`, this _will_ still be compiled so it also needs `#ifdef TOOLS_ENABLED` guards.

Geometror replied:
Ok, moved it under the `#ifdef TOOLS_ENABLED` line.

akien-mga commented on modules/websocket/register_types.cpp:
```diff
@@ -43,7 +43,7 @@
 #include "editor/debugger/editor_debugger_server.h"
 #include "editor/editor_node.h"
-#include "editor_debugger_server_websocket.h"
+#include "modules/websocket/editor_debugger_server_websocket.h"
```
Should not hardcode absolute path in a module

akien-mga commented on servers/rendering/renderer_scene_cull.h:
```diff
@@ -33,13 +33,9 @@
-#include "core/math/geometry_3d.h"
-#include "core/math/octree.h"
-#include "core/os/semaphore.h"
-#include "core/os/thread.h"
```
Nitpick, but these two blocks could be merged.

akien-mga reviewed (approved):
Excellent work overall! I reviewed the whole diff and commented on some style issues and a couple problems, but that's a big improvement already as is.

Geometror committed:
Improve compilation speed (forward declarations/includes cleanup)

Geometror commented:
Fixed all issues, except the ones regarding the dependency editor header as suggested. I was tempted to fix those, but there are still many places left where a similar cleanup could be done so this is indeed material for another PR :)

akien-mga commented:
Thanks a lot! I know this was tedious work but it's really appreciated :)
